src/workerd/jsg/async-context.h

```diff
@@ -0,0 +1,178 @@
+  // is created, it inherits a copy storage context of the parent.
+public:
+  class StorageKey: public kj::Refcounted {
+    // An opaque key that identifies an async-local storage cell within the frame.
+  public:
+    StorageKey() : hash(kj::hashCode(this)) {}
+
+    void reset() { dead = true; }
+    // The owner of the key should reset it when it goes away.
+    // The StorageKey is typically owned by an instance of AsyncLocalstorage (see
+    // the api/node/async-hooks.h). When the ALS instance is garbage collected, it
+    // must call reset to signal that this StorageKey is "dead" and can never be
+    // looked up again. Subsequent accesses to a frame will remove dead keys from
+    // the frame lazily. The lazy cleanup does mean that values may persist in
+    // memory a bit longer so if it proves to be problematic we can make the cleanup
+    // a bit more proactive.
+    //
+    // TODO(later): We should also evaluate the relatively unlikely case where an
+    // ALS is capturing a reference to itself and therefore can never be cleaned up.
+
+    bool isDead() const { return dead; }
+    inline uint hashCode() const { return hash; }
+    inline bool operator==(const StorageKey& other) const {
+      return hash == other.hash;
+    }
+
+  private:
+    uint hash;
+    bool dead = false;
+  };
+
+  struct StorageEntry {
+    kj::Own<StorageKey> key;
+    Value value;
+
+    inline StorageEntry clone(Lock& js) {
+      return {
+        .key = addRef(*key),
+        .value = value.addRef(js)
+      };
+    }
+  };
+
+  AsyncContextFrame(IsolateBase& isolate);
+  AsyncContextFrame(
+      Lock& js,
+      kj::Maybe<AsyncContextFrame&> maybeParent = nullptr,
+      kj::Maybe<StorageEntry> maybeStorageEntry = nullptr);
+
+  static AsyncContextFrame& current(Lock& js);
+  // Returns the reference to the AsyncContextFrame currently at the top of the stack.
+
+  static kj::Own<AsyncContextFrame> create(
+      Lock& js,
+      kj::Maybe<AsyncContextFrame&> maybeParent = nullptr,
+      kj::Maybe<StorageEntry> maybeStorageEntry = nullptr);
+  // Create a new AsyncContextFrame. If maybeParent is not specified, uses the current().
+  // If maybeStorageEntry is non-null, the associated storage cell in the new frame is
+  // set to the given value.
+
+  static v8::Local<v8::Function> attachContext(Lock& js, v8::Local<v8::Function> fn,
+                                      kj::Maybe<AsyncContextFrame&> maybeFrame = nullptr,
+                                      kj::Maybe<v8::Local<v8::Value>> thisArg = nullptr);
+  // Associates the given JavaScript function with the given AsyncContextFrame, returning
+  // a wrapper function that will ensure appropriate propagation of the async context
+  // when the wrapper function is called. If maybeFrame is not specified, the current()
+  // frame is used.
+
+  static v8::Local<v8::Promise> attachContext(Lock& js, v8::Local<v8::Promise> promise,
```

Comment:
Why does `attachContext()` return the promise? It's always the same handle passed in, right? Can it return void?